stdlib/public/Concurrency/Task.swift

```diff
@@ -623,6 +632,81 @@ extension Task where Success == Never, Failure == Never {
       _enqueueJobGlobalWithDelay(duration, job)
     }
   }
+
+  /// Suspends the current task for _at least_ the given duration
+  /// in nanoseconds, unless the task is cancelled. If the task is cancelled,
+  /// throws \c CancellationError.
+  ///
+  /// This function does _not_ block the underlying thread.
+  public static func sleep(nanoseconds duration: UInt64) async throws {
+    // Allocate a word to store a flag indicating that someone has picked
+    // up the continuation, and initialize it to zero.
+    let wordPtr = UnsafeMutablePointer<Builtin.Word>.allocate(capacity: 1)
+    Builtin.atomicstore_seqcst_Word(
+        wordPtr._rawValue, UInt(0)._builtinWordValue)
+
+    // When we leave, remove the cancellation record that's used to detect
+    // when the task has been cancelled.
+    var cancellationRecord: UnsafeRawPointer? = nil
+    defer {
+      _taskRemoveCancellationHandler(record: cancellationRecord!)
+    }
+
+    let _: () = try await withUnsafeThrowingContinuation { continuation in
+      // Cancellation handler function that executes if the current task is
+      // cancelled.
+      @Sendable func onCancel() {
+        // Indicate that we've finished by putting a "2" into the flag word.
+        let (_, won) = Builtin.cmpxchg_seqcst_seqcst_Word(
+            wordPtr._rawValue,
+            UInt(0)._builtinWordValue,
+            UInt(2)._builtinWordValue)
+
+        if Bool(_builtinBooleanLiteral: won) {
+          // We won because the task was cancelled before the sleep finished,
+          // so invoke the continuation with a cancellation error.
+          continuation.resume(throwing: _Concurrency.CancellationError())
+        } else {
+          // The sleep finished first, which means the the continuation was
+          // called by the operation. All we have left to do is deallocate the
+          // flag word storage.
+          wordPtr.deallocate()
```

Comment:
Does this leak `wordPtr` if the task isn’t cancelled?

Reply:
It does! I've fixed this in my slightly-less-broken-but-still-somehow-wrong version.